### PR TITLE
[wifi_info_flutter] Check Permissions in Android O or higher

### PR DESCRIPTION
**Pull request by hamdikahloun**
## Description

* Android: Add Log warning for unsatisfied requirement(s) in Android O or higher.
* Android: Update Example project.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by stuartmorgan** [comment]
@bparrishMines Can you take a look to evaluate whether this is something we need to take in the short term?

@hamdikahloun For the longer term, you should make sure to submit this to the https://plus.fluttercommunity.dev/ version of the plugin if applicable there, as we hope in time to be able to transition users to those versions of the plugins.

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 1]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+              + "For more information about Wi-Fi Restrictions in Android 8.0 and above, please consult the following link:\n"
+              + "https://developer.android.com/guide/topics/connectivity/wifi-scan");
+      return false;
+    }
+
+    if (Build.VERSION.SDK_INT == Build.VERSION_CODES.P && !grantedChangeWifiState) {
+      Log.w(
+          TAG,
+          "Attempted to get Wi-Fi data that requires additional permission(s).\n"
+              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android O, please ensure your app has the CHANGE_WIFI_STATE permission.\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:110]: Should this be `P`?
```suggestion
              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android P, please ensure your app has the CHANGE_WIFI_STATE permission.\n"
```

**Review by bparrishMines — COMMENTED** [review]
@stuartmorgan Yea, I think it makes sense to include this to improve the current error handling of the plugin. I did a review and left a comment.

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 2]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+              + "https://developer.android.com/guide/topics/connectivity/wifi-scan");
+      return false;
+    }
+
+    if (Build.VERSION.SDK_INT == Build.VERSION_CODES.P && !grantedChangeWifiState) {
+      Log.w(
+          TAG,
+          "Attempted to get Wi-Fi data that requires additional permission(s).\n"
+              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android P, please ensure your app has the CHANGE_WIFI_STATE permission.\n"
+              + "For more information about Wi-Fi Restrictions in Android 8.0 and above, please consult the following link:\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:111]:
```suggestion
              + "For more information about Wi-Fi Restrictions in Android 9.0 and above, please consult the following link:\n"
```

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 3]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+    if (Build.VERSION.SDK_INT == Build.VERSION_CODES.P
+        && !grantedAccessFine
+        && !grantedAccessCoarse) {
+      Log.w(
+          TAG,
+          "Attempted to get Wi-Fi data that requires additional permission(s).\n"
+              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android O, additional to CHANGE_WIFI_STATE please ensure your app has one of the following permissions too:\n"
+              + "- ACCESS_FINE_LOCATION\n"
+              + "- ACCESS_COARSE_LOCATION\n"
+              + "For more information about Wi-Fi Restrictions in Android 8.0 and above, please consult the following link:\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:125]:
```suggestion
              + "For more information about Wi-Fi Restrictions in Android 9.0 and above, please consult the following link:\n"
```

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 4]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+      return false;
+    }
+
+    if (Build.VERSION.SDK_INT == Build.VERSION_CODES.P
+        && !grantedAccessFine
+        && !grantedAccessCoarse) {
+      Log.w(
+          TAG,
+          "Attempted to get Wi-Fi data that requires additional permission(s).\n"
+              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android O, additional to CHANGE_WIFI_STATE please ensure your app has one of the following permissions too:\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:122]:
```suggestion
              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android P, additional to CHANGE_WIFI_STATE please ensure your app has one of the following permissions too:\n"
```

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 5]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+              + "https://developer.android.com/guide/topics/connectivity/wifi-scan");
+      return false;
+    }
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
+        && (!grantedAccessFine || !grantedChangeWifiState)) {
+      Log.w(
+          TAG,
+          "Attempted to get Wi-Fi data that requires additional permission(s).\n"
+              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android O, please ensure your app has the CHANGE_WIFI_STATE and ACCESS_FINE_LOCATION permission.\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:135]:
```suggestion
              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android Q, please ensure your app has the CHANGE_WIFI_STATE and ACCESS_FINE_LOCATION permission.\n"
```

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 6]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+              + "For more information about Wi-Fi Restrictions in Android 8.0 and above, please consult the following link:\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:136]:
```suggestion
              + "For more information about Wi-Fi Restrictions in Android 10.0 and above, please consult the following link:\n"
```

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 7]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+    LocationManager locationManager =
+        (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
+
+    boolean gpsEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && !gpsEnabled) {
+      Log.w(
+          TAG,
+          "Attempted to get Wi-Fi data that requires additional permission(s).\n"
+              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android O, please ensure Location services are enabled on the device (under Settings > Location).\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:150]:
```suggestion
              + "To successfully get WiFi Name or Wi-Fi BSSID starting with Android P, please ensure Location services are enabled on the device (under Settings > Location).\n"
```

File: packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java [hunk 8]
```diff
@@ -53,4 +69,89 @@ String getWifiIPAddress() {
...
+              + "For more information about Wi-Fi Restrictions in Android 8.0 and above, please consult the following link:\n"
```
Comment (bparrishMines) [inline packages/wifi_info_flutter/wifi_info_flutter/android/src/main/java/io/flutter/plugins/wifi_info_flutter/WifiInfoFlutter.java:151]:
```suggestion
              + "For more information about Wi-Fi Restrictions in Android 9.0 and above, please consult the following link:\n"
```

**Review by bparrishMines — APPROVED** [review]
LGTM

Everything else looks good. I just noticed a few more potential fixes for the versions in the logging.

**Comment by hamdikahloun** [comment]
Thank you, PR updated.

**Pull request merged**